2021-01-20-reproducible-progress.md

```diff
@@ -0,0 +1,81 @@
+- More tests, specifically for reproducible builds of packages
+
+We also have a nice [summary page] with all the recent jobs states, which is updated daily.
+
+
+Reproducible builds
+--------------------
+
+Another major work item is progressing [reproducible builds] in Qubes OS. Our current short-term goal is to integrate what is realistically possible in Debian. Specifically, we aim to get our Debian templates to the state in which packages can be installed only when configured rebuilders confirm they really came from the source code we publish. For this task, we focus specifically on Debian Bullseye (testing version), which includes most of the reproducibility work done by the Debian community.
+
+To start, we had to bring our packages to the state in which they could be built reproducibly. GitLab-CI massively helped here with identifying issues. One of the CI jobs we added is running the `reprotest` tool on our packages. This tool builds the package twice in two different build environments and compares the results. If there are any differences, the job automatically runs the `diffoscope` tool, which provides a detailed, in-depth report.
+Now, all our packages (for Debian Bullseye) are reproducible, and we are actively monitoring it with our CI. This can be seen in the "reproducible" badges in the CI [summary page].
+
+The next step on the journey is enabling an arbitrary user or organization to challenge the official packages --- to take the sources, together with the `buildinfo` file describing how it was built and verify that they result in the same binary package. Theoretically, there is a [`debrebuild.pl`] script in Debian's `devscripts` package that does exactly this. We started by adding missing parts to it, but it turned out there were several limitations making it unsuitable for our use case. The primary limitation is support for only one source repository. (We need two: original Debian and Qubes.) Since `debrebuild.pl`, as the name suggests, is written in Perl, it isn't very friendly for contributing to. For these reasons, we decided to write our own similar tool in Python. While this was a bit of work, [the end result][debrebuild] is very nice: about 900 lines of code, with proper structure using classes. We are in the process of upstreaming this code back to Debian, potentially to replace the original `debrebuild.pl` tool.
+
+To ease the rebuilder setup, we are also in the process of creating a simple [rebuilder orchestrator] that monitors specified repositories and does the package build verification automatically when necessary. The goal is to give anyone who wants to verify our (or others) packages an easy way to get started. The decision of who to trust is separate, but in order to have a choice, rebuilders must exist in the first place. This also includes scenarios in which users or organizations want to verify packages for themselves. 
```

Comment:
> The decision of who to trust is separate, but in order to have a choice, rebuilders must exist in the first place.

This sentence is somewhat mysterious. I suggest adding more context explaining what you have in mind.

Reply:
Maybe something like: "The goal is to give anyone who wants to verify our (or others) packages an easy way to get started. Of course, they have to choose which rebuilder to trust (and this is a separate choice) - but there must exist some rebuilders to choose from in the first place."

Although I would really welcome a definition of a rebuilder here, tbh I'm not sure what it really means.

Reply:
Rebuilder is a thing that given a binary package and its source (plus some extra metadata if applicable), tries to build that package again (rebuild), to verify if that binary really comes from the source it claims to use.